cpp/src/dataframe/device_table.cuh

```diff
@@ -0,0 +1,331 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef DEVICE_TABLE_H
+#define DEVICE_TABLE_H
+
+#include <cudf.h>
+#include <rmm/thrust_rmm_allocator.h>
+#include <utilities/cudf_utils.h>
+#include <bitmask/legacy_bitmask.hpp>
+#include <hash/hash_functions.cuh>
+#include <hash/managed.cuh>
+#include <types.hpp>
+#include <utilities/error_utils.hpp>
+#include <utilities/type_dispatcher.hpp>
+
+#include <thrust/iterator/counting_iterator.h>
+#include <thrust/logical.h>
+#include <thrust/tabulate.h>
+
+/**
+ * @brief Lightweight wrapper for a device array of `gdf_column`s of the same
+ * size
+ *
+ */
+class device_table {
+ public:
+  /**---------------------------------------------------------------------------*
+   * @brief Factory function to construct a device_table wrapped in a
+   * unique_ptr.
+   *
+   * Because this class allocates device memory, and we want to be sure that
+   * device memory is free'd, this factory function returns a device_table
+   * wrapped in a unique_ptr with a custom deleter that frees the device memory.
+   *
+   * The class' destructor does **not** free the device memory, because we would
+   * like to be able to pass instances of this class by value into kernels via a
+   * shallow-copy (i.e., just copying the pointers without allocating any new
+   * device memory). Since it is shallow copied, if the destructor were to free
+   * the device memory, then you would end up trying to free the underlying
+   * device memory any time a copy is destroyed.
+   *
+   * Instead, the underlying device memory will not be free'd until the returned
+   * `unique_ptr` invokes its deleter.
+   *
+   * Usage:
+   * ```
+   * gdf_column * col;
+   * auto device_table_ptr = device_table::create(1, &col);
+   *
+   * // Table is passed by **value**, i.e., shallow copy into kernel
+   * some_kernel<<<...>>>(*device_table_ptr);
+   * ```
+   *
```

Comment:
Should you comment that the stream is not synchronized / this function is async when `stream` is nonzero?

Reply:
Usually I assume that any function that takes a stream argument will be asynchronous wrt stream ordering and not artificially synchronize that stream.

Reply:
So you are saying that's not worth a single line in the documentation? Once we gain some consistency, we can move such documentation to a top level that applies to all APIs.